### PR TITLE
[AgriFood] Adding Nov 2022 API version update with STAC Feature GET and Search APIs

### LINKED ISSUE #22336: [AgriFood Platform] API Review
New API Review meeting has been requested.

**Service Name**: Azure AgriFood Platform
**Review Created By**: Bhanu Prakash Chinthalapally @bhanuprakash-1 
**Review Date**: 2/9/2023 4:00 PM PT
**PR**: 
**Hero Scenarios Link**: Not Provided
**Core Concepts Doc Link**: Not Provided

**Description**: Adding two new DataPlane APIs for our service.

Detailed meeting information and documents provided can be accessed [here](https://web.powerapps.com/apps/e41c69de-ef96-44ba-b287-0e012e7c13ae?view=edit&event=12f7cec0-459e-ed11-aad1-000d3a3418aa)
For more information that will help prepare you for this review, the requirements, and office hours, visit the documentation [here](https://aka.ms/azsdk/onboarding/restapischedule)

### PR DESCRIPTION
**Pull request by bhanuprakash-1**
# Data Plane API - Pull Request
Adding two new APIs for STAC (Spatio Temporal Asset Catalog) Feature get and Search features Functionality. 
Get feature:
GET /scenes/collections/{collectionId}/features/{featureId}
this will accept an input of: collectionId, featureId and returns STAC Feature in the given collection Id.

Search feature:
 POST /scenes/collections/{collectionId}:search
Input: geometry (bbox or polygon, multipolygon), start and end datetime and collectionId, (optional array) feature_id(s)(size of array [0,5]), limit (size of returned list) and next (next token to get next set of results).
Ouput: List of Features, next token and returned no of features.

Weather API Changes:
Following up on (this PR)[[[AgriFood] Adding Feb 2023 API version by JatinKhatri03 · Pull Request #11131 · Azure/azure-rest-api-specs-pr (github.com)](https://github.com/Azure/azure-rest-api-specs-pr/pull/11131)]. Instead of bumping up the api version, we are going to add these changes in current api-version(1 and 2 are breaking changes) for "/weather-data/:fetch" api 1) Making all the WeatherDataForPassthrough fields an array to support more compact response (this helped us to reduce the data size by more than 50%).
2) Adding Measures model that will now support a "value" array.
3) Renamed Locations to WeatherLocationData.
4) Mentioned "x-ms-cache-hit" and "x-ms-response-latency" in the swagger.

##Other Spec Updates:
1) Adding new satellite source Sentinel_2_L1C.
2) Changing CropVariety to CropProduct.
3) Adding errorCode in jobs response body.
4) Updating Weather Data example and method name.

## API Info: The Basics
Most of the information about your service should be captured in the issue that serves as your [*engagement record*](https://dev.azure.com/azure-sdk/internal/_wiki/wikis/internal.wiki/271/Azure-REST-API-Stewardship?anchor=rest-api-stewardship-process).

* Link to engagement record issue: https://github.com/Azure/azure-rest-api-specs/issues/22336

Is this review for (select one):

- [ ] a private preview
- [ ] a public preview
- [ ] GA release

### Change Scope
<sup>This section will help us focus on the specific parts of your API that are new or have been modified. <br/>Please share a link to the design document for the new APIs, a link to the previous Open API document (swagger) if applicable, and the root paths that have been updated. </sup>
* Design Document:
* Previous Open API Doc:
* Updated paths:

## ❔Got questions? Need additional info?? We are here to help!

<details>
  <summary> Contact us!</summary>

The [Azure API Review Board](https://aka.ms/azapi) is dedicated to helping you create amazing APIs. You can read about our mission and learn more about our process on our [wiki](https://aka.ms/azapi).
* 💬 [Teams Channel](https://teams.microsoft.com/l/channel/19%3a3ebb18fded0e47938f998e196a52952f%40thread.tacv2/General?groupId=1a10b50c-e870-4fe0-8483-bf5542a8d2d8&tenantId=72f988bf-86f1-41af-91ab-2d7cd011db47)
* 💌 [email](mailto://azureapirbcore@microsoft.com)

</details>

<details>
  <summary>Click here for links to tools, specs, guidelines & other good stuff</summary>

### Tooling
 * [Open API validation tools](https://aka.ms/swaggertools) were run on this PR. Go here to see [how to fix errors](https://aka.ms/ci-fix)
 * [Spectral Linting](https://aka.ms/style)
 * [Open API Hub](https://aka.ms/openapiportal)

### Guidelines & Specifications
 * [Azure REST API Guidelines](https://aka.ms/guidelines)
 * [OpenAPI Style Guidelines](https://aka.ms/style)
 * [Azure Breaking Change Policy](https://aka.ms/AzBreakingChangesPolicy)

### Helpful Links
 * [Azure DevTools Wiki](https://aka.ms/azapi)

</details>

### CODE REVIEW

File: specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/agfood.json [hunk 1]
```diff
@@ -12716,9 +12771,9 @@
           {
             "in": "query",
             "name": "source",
-            "description": "Source name of scene data, default value Sentinel_2_L2A (Sentinel 2 L2A).",
-            "type": "string",
-            "default": "Sentinel_2_L2A"
+            "description": "Source name of scene data, Available Values: Sentinel_2_L2A, Sentinel_2_L1C.",
+            "required": true,
+            "type": "string"
```
Comment (weidongxu-microsoft) [inline specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/agfood.json:12776]: prefer extensible enum https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#enums--sdks-client-libraries

File: specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/agfood.json [hunk 2]
```diff
@@ -12840,6 +12895,126 @@
         }
       }
     },
+    "/scenes/collections/{collectionId}/features/{featureId}": {
+      "get": {
+        "tags": [
+          "Scenes"
+        ],
+        "description": "Get a feature(STAC Item) for given collection and feature id.",
```
Comment (weidongxu-microsoft) [inline specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/agfood.json:12903]: Can we also include the name of STAC in the description? 

E.g. SpatioTemporal Asset Catalog (STAC) Item

File: specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/agfood.json [hunk 3]
```diff
@@ -12840,6 +12895,126 @@
...
+        "produces": [
+          "application/json"
+        ],
+        "parameters": [
+          {
+            "in": "path",
+            "name": "collectionId",
+            "description": "Collection Id to be fetched. Available Values are: Sentinel_2_L2A, Sentinel_2_L1C.",
+            "required": true,
+            "type": "string"
```
Comment (weidongxu-microsoft) [inline specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/agfood.json:12914]: similar comment, prefer extensible enum

File: specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/agfood.json [hunk 4]
```diff
@@ -12840,6 +12895,126 @@
...
+          }
+        },
+        "x-ms-examples": {
+          "Scenes_GetStacFeature": {
+            "$ref": "./examples/Scenes_GetStacFeature.json"
+          }
+        }
+      }
+    },
+    "/scenes/collections/search": {
```
Comment (weidongxu-microsoft) [inline specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/agfood.json:12959]: please check if this is an action https://github.com/microsoft/api-guidelines/blob/vNext/azure/Guidelines.md#performing-an-action

may use `/scenes/collections:search` or `/scenes:search`

File: specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/agfood.json [hunk 5]
```diff
@@ -23507,19 +23811,150 @@
...
+          "description": "Bounding box of the feature.",
+          "type": "array",
+          "items": {
+            "format": "double",
+            "type": "number"
+          }
+        },
+        "properties": {
+          "description": "A dictionary of additional metadata for the item."
+        },
```
Comment (weidongxu-microsoft) [inline specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/agfood.json:23876]: type: object ?

File: specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/agfood.json [hunk 6]
```diff
@@ -23507,19 +23811,150 @@
...
+          "type": "string"
+        },
+        "type": {
+          "description": "Type of the GeoJSON Object. It's value is always Feature.",
+          "minLength": 1,
+          "type": "string"
+        },
+        "geometry": {
+          "description": "Defines the full footprint of the asset represented by this item.\r\nIts a GeoJSON geometry."
+        },
```
Comment (weidongxu-microsoft) [inline specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/agfood.json:23865]: type: object ?

File: specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/agfood.json [hunk 7]
```diff
@@ -23507,19 +23811,150 @@
...
+          }
+        },
+        "assets": {
+          "description": "Dictionary of asset objects, each with a unique key.",
+          "type": "object",
+          "additionalProperties": {
+            "$ref": "#/definitions/StacFeatureAsset"
+          }
+        },
+        "collection": {
```
Comment (weidongxu-microsoft) [inline specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/agfood.json:23891]: collectionId ?

File: specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/agfood.json [hunk 8]
```diff
@@ -23632,7 +24067,8 @@
             "key2": 123.45
           }
         }
-      }
+      },
+      "additionalProperties": {}
```
Comment (weidongxu-microsoft) [inline specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/agfood.json:24071]: I am not sure the purpose of this "additionalProperties"

File: specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/agfood.json (hunk 7 above)
Comment (bhanuprakash-1) [reply specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/agfood.json:23891]: This model's properties are according to an external specification, and it refers to this as "collection" only.

File: specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/agfood.json (hunk 3 above)
Comment (bhanuprakash-1) [reply specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/agfood.json:12914]: This is a path parameter. Not sure if we can have enums in path parameters. We use Swashbuckle.AspNetCore Cli Tool and its not generating the swagger properly if we have enums in path parameters, but works fine if its in POST/GET body.

**Comment by weidongxu-microsoft** [comment]
I've re-run the CI. Please check SemanticValidation and ModelValidation first.

File: specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/agfood.json (hunk 3 above)
Comment (weidongxu-microsoft) [reply specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/agfood.json:12914]: I think we can.

The problem might be Swashbuckle. Generally it is not recommended. If it is used, there likely to be some manual modification on the output (like this case).

**Comment by bhanuprakash-1** [comment]
> I've re-run the CI. Please check SemanticValidation and ModelValidation first.

I see the failures in json validation, but I don't find those semantic and model validation errors when checked with avocado locally. The error points to the first line itself, but unable to figure out what is the error. Check with external json validators also, didn't find any issue.

**Comment by weidongxu-microsoft** [comment]
@raych1 
Hi Ray, do you know what's the reason of SemanticValidation?

I seem not able to read useful information from logs.

File: specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/agfood.json (hunk 1 above)
Comment (bhanuprakash-1) [reply specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/agfood.json:12776]: Facing same issue as with path params. Unable to use enums in query parameters. We are using swashbuckle to generate swagger and when using enums in query, type info is not getting populated.

File: specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/agfood.json (hunk 6 above)
Comment (bhanuprakash-1) [reply specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/agfood.json]: Done.

File: specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/agfood.json (hunk 8 above)
Comment (bhanuprakash-1) [reply specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/agfood.json]: Removed.

File: specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/agfood.json (hunk 5 above)
Comment (bhanuprakash-1) [reply specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/agfood.json]: Done.

**Comment by weidongxu-microsoft** [comment]
> @raych1 Hi Ray, do you know what's the reason of this SemanticValidation on 1:1?
> 
> I seem not able to read useful information from logs.

@bhanuprakash-1
Ray mentioned it could be [BOM](https://en.wikipedia.org/wiki/Byte_order_mark) in your JSON. You might want to open it in editor and see if it got option to encode to UTF without BOM.

File: specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/agfood.json (hunk 2 above)
Comment (bhanuprakash-1) [reply specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/agfood.json]: Added the whole name.

**Comment by bhanuprakash-1** [comment]
> > @raych1 Hi Ray, do you know what's the reason of this SemanticValidation on 1:1?
> > I seem not able to read useful information from logs.
> 
> @bhanuprakash-1 Ray mentioned it could be [BOM](https://en.wikipedia.org/wiki/Byte_order_mark) in your JSON. You might want to open it in editor and see if it got option to encode to UTF without BOM.

Thanks @weidongxu-microsoft , errors are gone after changing the encoding from UTF+BOM  to UTF-8. Not sure how BOM cropped in.

File: specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/examples/Scenes_SearchFeatures.json [hunk 9]
```diff
@@ -433,8 +433,7 @@
             }
           }
         ],
-        "next": 17,
-        "returned": 2
+        "nextLink": "https://{{resourceName}}.farmbeats.azure.net//scenes/stac-collections/{collectionId}:search?api-version={{apiVersion}}&collectionId=Sentinel_2_L2A&maxpagesize=2&skip=17"
```
Comment (weidongxu-microsoft) [inline specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/examples/Scenes_SearchFeatures.json:436]: `collectionId=Sentinel_2_L2A` part seems wrong?

`collectionId` should not be in query parameter, but in path e.g. `https://{{resourceName}}.farmbeats.azure.net//scenes/stac-collections/Sentinel_2_L2A:search?...`

I know this is an example, but make sure your backend is correct.

Comment (bhanuprakash-1) [reply specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/examples/Scenes_SearchFeatures.json]: It was manual error in example. Updated it. Thanks!

**Comment by bhanuprakash-1** [comment]
Hi @weidongxu-microsoft , can you please approve the PR?

We have noted down the below suggestions for future versions:
1) agrifood currently return STAC data directly, do we recommend them keep the exact STAC specs, i.e. “stac_version”; or use Azure camelCase, i.e. “stacVersion”; or leave it as a future discussion, as I understand that backend  is still private preview?

2) Establish a plan now to handle new versions of the STAC, what is the back-up plan when third party providers do breaking API schema changes?

3) Please check the URL for the action pattern should be/weather-data:fetch .

File: specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/agfood.json [hunk 10]
```diff
@@ -13018,6 +13045,168 @@
...
+            "required": true,
+            "type": "string",
+            "enum": [
+              "Sentinel_2_L2A",
+              "Sentinel_2_L1C"
+            ],
+            "x-ms-enum": {
+              "name": "collectionId",
+              "modelAsString": true
+            }
```
Comment (weidongxu-microsoft) [inline specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/agfood.json:13075]: One more issue that you can resolve in this PR or in future api-version (before you request SDK, this won't affect REST API):

Unify the name of the extensible enum.
Here it is `"name": "Source"`
https://github.com/Azure/azure-rest-api-specs/blob/e38daec67d57ef9c4804b1e3055753407e45fa71/specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/agfood.json#L18473-L18481

Though we call it extensible enum, it more akin to "string with known values" that has a "name" in SDK. And here we'd like same extensible enum having same name.

**Review by weidongxu-microsoft — APPROVED** [review]

**Comment by weidongxu-microsoft** [comment]
More more "TODO" is the "WeatherDataForPassthrough" by Jatin.

- Recommend to separate the processed data, and the raw data from provider. (down-side seems that this might increase payload size)

File: specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/agfood.json (hunk 10 above)
Comment (bhanuprakash-1) [reply specification/agrifood/data-plane/Microsoft.AgFoodPlatform/preview/2022-11-01-preview/agfood.json:13075]: Sure. Will update in future versions. Although the values are same now, they can diverge in future. Hence kept different names for future flexibility.

**Pull request merged**